src/Orleans/Utils/IKeyedServiceDescriptor.cs

```diff
@@ -0,0 +1,85 @@
+﻿
+using System;
+using System.Linq;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Orleans
+{
+    public static class KeyedServiceDescriptorExtensions
+    {
+        /// <summary>
+        /// Add supported for keyed services to a service collection
+        /// </summary>
+        public static IServiceCollection SupportKeyedServices(this IServiceCollection collection)
+        {
+            return collection.AddSingleton(typeof(IKeyedServiceCollection<,>), typeof(KeyedServiceCollection<,>));
+        }
+
+        /// <summary>
+        /// Associates a key with a specialization of a service type.
+        /// </summary>
+        public static IServiceCollection WithKey<TKey, TService, TSpecialization>(this IServiceCollection collection, TKey key)
+            where TService : class
+            where TSpecialization : class, TService
+        {
+            return collection.AddSingleton<IKeyedServiceDescriptor<TKey, TService>>(sp => new KeyedServiceDescriptor<TKey, TService, TSpecialization>(key));
+        }
+
+        /// <summary>
+        /// Associates a name with a specialization of a service type.
+        /// </summary>
+        public static IServiceCollection WithName<TService, TSpecialization>(this IServiceCollection collection, string name)
+            where TService : class
+            where TSpecialization : class, TService
+        {
+            return collection.WithKey<string, TService, TSpecialization>(name);
+        }
+    }
+
+    internal interface IKeyedServiceDescriptor<TKey, TService> : IEquatable<TKey>
+        where TService : class
+    {
+        Type Service { get; }
+    }
+
+    internal class KeyedServiceDescriptor<TKey, TService, TInstance> : IKeyedServiceDescriptor<TKey, TService>
+            where TService : class
+            where TInstance : TService
+    {
+        private readonly TKey key;
+
+        public KeyedServiceDescriptor(TKey key)
+        {
+            this.key = key;
+            this.Service = typeof(TInstance);
+        }
+
+        public Type Service { get; }
+
+        public bool Equals(TKey other)
+        {
+            return Equals(key, other);
+        }
+    }
+
+    internal class KeyedServiceCollection<TKey, TService> : IKeyedServiceCollection<TKey, TService>
+        where TService : class
+    {
+        private readonly IKeyedServiceDescriptor<TKey, TService>[] keyedServices;
+        private readonly IServiceProvider services;
+
+        public KeyedServiceCollection(IServiceProvider services)
+        {
+            this.keyedServices = services.GetServices<IKeyedServiceDescriptor<TKey, TService>>().ToArray();
+            this.services = services;
+        }
+
+        public TService GetService(TKey key)
+        {
+            IKeyedServiceDescriptor<TKey, TService> keyedService = this.keyedServices.FirstOrDefault(s => s.Equals(key));
```

Comment:
Correctness here would be more obvious if `TKey` was exposed as an accessible property and used for the comparison rather than implementing `IEquatable<TKey>`.